blogposts/2020/optimizing-a-code-intelligence-indexer.md

```diff
@@ -0,0 +1,196 @@
+
+##### You have to be sure that there are no hidden data dependencies
+
+When we index references, we assume that our map of definitions is already populated. Our solution is to parallelize the workload in distinct steps (first index all definitions, then index all references), where each step can rely on data calculated by hte previous step which was parallelized individually.
+
+Alternative solutions here require pushing back tasks whose dependencies have not yet been met, pre-determining a dependency graph, or recursively submitting duplicate work to the queue. Each of these solutions have a down side and decrease code clarity.
+
+##### You have to choose the right level of work to parallelize
+
+You can't throw work into a parallel executor and guarantee that your program runs faster. In fact, the first few attempts at parallelizing lsif-go wasn't fruitful because we tried to index each _file_ in parallel rather than each _package_. The cost of indexing a single file was low enough that the overhead of setting up the goroutines and queueing the work negated any benefit of executing the work in multiple cores.
+
+The cost of indexing an entire package is high enough that there is obvious, tangible benefit of indexing multiple packages in different cores. This benefit far outweighs the any of the setup and queueing overhead.
+
+<div class="no-shadow">
+  <img src="https://sourcegraphstatic.com/blog/lsif-go/lsif-go-parallel.svg" alt="Parallelism diagram">
+</div>
+
+In lsif-go v1.0.0, all writes to the file are synchronized by a single goroutine controlling the output buffer. This ensures that for any single goroutine emitting elements, the order of the elements they emit will hit disk in the same order. This is important as LSIF requires that edges refer only to a vertex that has already been indexed.r
+
+Each non-trivial task is broken into package-level work, queued into a channel, and then distributed over a number of goroutines (that scales depending on the number of physical cores on the indexing machine). A task finishes once the entire input channel has been consumed and each worker goroutines has exited. Each task makes a fresh set of goroutines.
+
+## Reviewing results
+
+The following charts show the comparison between the last three versions of the indexer: v0.9.0, v0.10.0, and v1.0.0.
+
+<div class="no-shadow">
```

Comment:
What's the reason for showing three charts here? Doesn't the first chart show a superset of what the other 2 charts show?

Reply:
Yes but if you want to actually see the performance difference we'd need to zoom in to the other ones. The third might be overkill as it doesn't show a comparison, just performance, but I think the first two show a nice "we quadruped performance between 0.9 and 0.10" and a "we quadrupled performance again between 0.10 and 1.0".

Reply:
I'll update it to reflect that we don't care about tagged versions.